zebrad/src/components/sync.rs

```diff
@@ -16,38 +17,37 @@ use zebra_chain::{
 use zebra_network as zn;
 use zebra_state as zs;
 
+mod downloads;
+use downloads::{AlwaysHedge, Downloads};
+
 /// Controls the number of peers used for each ObtainTips and ExtendTips request.
 const FANOUT: usize = 4;
 
 /// Controls how many times we will retry each block download.
 ///
-/// If all the retries fail, then the syncer will reset, and start downloading
-/// blocks from the verified tip in the state, including blocks which previously
-/// downloaded successfully.
-///
-/// But if a node is on a slow or unreliable network, sync restarts can result
-/// in a flood of download requests, making future syncs more likely to fail.
-/// So it's much faster to retry each block multiple times.
+/// Failing block downloads is important because it defends against peers who
+/// feed us bad hashes. But spurious failures of valid blocks cause the syncer to
+/// restart from the previous checkpoint, potentially re-downloading blocks.
 ///
-/// When we implement a peer reputation system, we can reduce the number of
-/// retries, because we will be more likely to choose a good peer.
-const BLOCK_DOWNLOAD_RETRY_LIMIT: usize = 5;
+/// We also hedge requests, so we may retry up to twice this many times.
+const BLOCK_DOWNLOAD_RETRY_LIMIT: usize = 2;
 
 /// Controls how far ahead of the chain tip the syncer tries to download before
-/// waiting for queued verifications to complete. Set to twice the maximum
-/// checkpoint distance.
+/// waiting for queued verifications to complete.
 ///
-/// Some checkpoints contain larger blocks, so the maximum checkpoint gap can
-/// represent multiple gigabytes of data.
+/// Increasing this limit increases the buffer size, so it reduces the impact of
+/// missing a block on the critical path. The block size limit is 2MB, so in
+/// theory, this could represent multiple gigabytes of data, if we downloaded
+/// arbitrary blocks. However, because we randomly load balance outbound
+/// requests, and separate block download from obtaining block hashes, an
+/// adversary would have to control a significant fraction of our peers to lead
+/// us astray.
 const LOOKAHEAD_LIMIT: usize = zebra_consensus::MAX_CHECKPOINT_HEIGHT_GAP * 2;
 
 /// Controls how long we wait for a tips response to return.
-///
-/// The network layer also imposes a timeout on requests.
 const TIPS_RESPONSE_TIMEOUT: Duration = Duration::from_secs(6);
+
 /// Controls how long we wait for a block download request to complete.
-///
-/// The network layer also imposes a timeout on requests.
 const BLOCK_DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(20);
```

Comment:
Since Hedge doesn't support timeouts 10 seconds or greater, should we reduce the block download timeout?
```suggestion
const BLOCK_DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(9);
```

Reply:
I'd rather that we keep this timeout to the correct value, and just fix the hedge issue upstream -- it's not that processing the request will fail if the request takes longer than 10 seconds, but just that recording the statistics will.

Reply:
Which means that more requests are hedged than necessary, because the 95th percentile is lower. I guess that's ok.